src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -0,0 +1,246 @@
+		 * output directly to COPY.
+		 */
+		if (columnType != selectOutputType)
+		{
+			Expr *selectExpression = selectTargetEntry->expr;
+			Expr *typeCastedSelectExpr =
+				(Expr *) coerce_to_target_type(NULL, (Node *) selectExpression,
+											   selectOutputType, columnType,
+											   columnTypeMod, COERCION_EXPLICIT,
+											   COERCE_IMPLICIT_CAST, -1);
+
+			selectTargetEntry->expr = typeCastedSelectExpr;
+		}
+	}
+
+	multiPlan->insertSelectSubquery = subquery;
+	multiPlan->insertTargetList = insertSelectQuery->targetList;
+	multiPlan->targetRelationId = targetRelationId;
+
+	return multiPlan;
+}
+
+
+/*
+ * ErrorIfCoordinatorInsertSelectUnsupported returns an error if executing an
+ * INSERT ... SELECT command by pulling results of the SELECT to the coordinator
+ * is unsupported because it uses RETURNING, ON CONFLICT, or an append-distributed
+ * table.
+ */
+static DeferredErrorMessage *
+DeferErrorIfCoordinatorInsertSelectUnsupported(Query *insertSelectQuery)
+{
+	RangeTblEntry *insertRte = NULL;
+	RangeTblEntry *subqueryRte = NULL;
+	Query *subquery = NULL;
+
+	if (list_length(insertSelectQuery->returningList) > 0)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "RETURNING is not supported in INSERT ... SELECT via "
+							 "coordinator", NULL, NULL);
+	}
+
+	if (insertSelectQuery->onConflict)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "ON CONFLICT is not supported in INSERT ... SELECT via "
+							 "coordinator", NULL, NULL);
+	}
+
+	insertRte = ExtractInsertRangeTableEntry(insertSelectQuery);
+	if (PartitionMethod(insertRte->relid) == DISTRIBUTE_BY_APPEND)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "INSERT ... SELECT into an append-distributed table is "
+							 "not supported", NULL, NULL);
+	}
+
+	subqueryRte = ExtractSelectRangeTableEntry(insertSelectQuery);
+	subquery = (Query *) subqueryRte->subquery;
+
+	if (NeedsDistributedPlanning(subquery) &&
+		contain_nextval_expression_walker((Node *) insertSelectQuery->targetList, NULL))
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "INSERT ... SELECT cannot generate sequence values when "
+							 "selecting from a distributed table",
+							 NULL, NULL);
+	}
+
+	return NULL;
+}
+
+
+/*
+ * DecrementCTELevelsWalker recursively lowers the ctelevelsup of all
+ * range table entries that refer to the top level CTE list in the
+ * given context. This needs to be done after moving the list of CTEs
+ * down one level.
+ */
+static bool
+DecrementCTELevelsWalker(Node *node, struct DecrementCTELevelsContext *context)
+{
+	bool walkIsComplete = false;
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, RangeTblEntry))
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) node;
+		if (rangeTableEntry->ctename != NULL &&
+			rangeTableEntry->ctelevelsup == context->topLevelsUp)
+		{
+			ListCell *cteCell = NULL;
+
+			foreach(cteCell, context->topLevelCTEList)
```

Comment:
Very minor -- feel free to overlook since the following comment might be personal:

 shall we use `insertCTEList` (or smth like that) instead of `topLevelCTEList`? Initially, at this point of the code, I became confused since this is a walker function and `topLevel` seemed top level of the recursion.